File: php/src/StdUriTemplate.php
<?php
namespace StdUriTemplate;

use InvalidArgumentException;
use TypeError;
use Exception;

class StdUriTemplate {

    /**
     * @param string $template
     * @param array $substitutions
     * @return string
     * @throws InvalidArgumentException
     */
    public static function expand(string $template, array $substitutions): string {
        return self::expandImpl($template, $substitutions);
    }

    /**
     * @param string $token
     * @param int $col
     * @throws InvalidArgumentException
     */
    private static function checkVarname(string $token, int $col): void {
        if ($token[0] === '.' || substr($token, -1) === '.') {
            throw new InvalidArgumentException("Invalid variable name at col: $col");
        }
        if (strpos($token, '..') !== false) {
            throw new InvalidArgumentException("Invalid variable name at col: $col");
        }
        $len = strlen($token);
        for ($i = 0; $i < $len; $i++) {
            if ($token[$i] === '%') {
                if ($i + 2 >= $len || !ctype_xdigit($token[$i + 1]) || !ctype_xdigit($token[$i + 2])) {
                    throw new InvalidArgumentException("Invalid variable name at col: $col");
                }
            }
        }
    }

    /**
     * @param string $c
     * @param int $col
     * @throws InvalidArgumentException
     */
    private static function validateLiteral(string $c, int $col): void {
        switch ($c) {
            case '+':
            case '#':
            case '/':
            case ';':
            case '?':
            case '&':
            case ' ':
            case '!':
            case '=':
            case '$':
            case '|':
            case '*':
            case ':':
            case '~':
            case '-':
                throw new InvalidArgumentException("Illegal character identified in the token at col: $col");
            default:
                break;
        }
    }

    /**
     * @param string|null $buffer
     * @param int $col
     * @return int
     * @throws InvalidArgumentException
     */
    private static function getMaxChar(?string $buffer, int $col): int {
        if ($buffer === null) {
            return -1;
        } else {
            $value = $buffer;

            if ($value === '') {
                return -1;
            } else {
                if (!ctype_digit($value)) {
                    throw new InvalidArgumentException("Cannot parse max chars at col: $col");
                }
                $intValue = (int)$value;
                if ($value[0] === '0') {
                    throw new InvalidArgumentException("Cannot parse max chars at col: $col");
                }
                if ($intValue < 1 || $intValue > 9999) {
                    throw new InvalidArgumentException("Cannot parse max chars at col: $col");
                }
                return $intValue;
            }
        }
    }

    /**
     * @param string $c
     * @param string &$token
     * @param int $col
     * @return string
     */
    private static function getOperator(string $c, string &$token, int $col): string {
        switch ($c) {
            case '+': return 'PLUS';
            case '#': return 'HASH';
            case '.': return 'DOT';
            case '/': return 'SLASH';
            case ';': return 'SEMICOLON';
            case '?': return 'QUESTION_MARK';
            case '&': return 'AMP';
            default:
                self::validateLiteral($c, $col);
                $token .= $c;
                return 'NO_OP';
        }
    }

    /**
     * @param string $str
     * @param array $substitutions
     * @return string
     * @throws InvalidArgumentException
     */
    private static function expandImpl(string $str, array $substitutions): string {
        $result = '';
        $token = null;

        $operator = null;
        $composite = false;
        $maxCharBuffer = null;
        $firstToken = true;

        for ($i = 0; $i < strlen($str); $i++) {
            $character = $str[$i];
            switch ($character) {
                case '{':
                    $token = '';
                    $firstToken = true;
                    break;
                case '}':
                    if ($token !== null) {
                        if ($maxCharBuffer !== null && $maxCharBuffer === '') {
                            throw new InvalidArgumentException("Found an empty prefix at col: $i");
                        }
                        $expanded = self::expandToken($operator, $token, $composite, self::getMaxChar($maxCharBuffer, $i), $firstToken, $substitutions, $result, $i);
                        if ($expanded && $firstToken) {
                            $firstToken = false;
                        }
                        $token = null;
                        $operator = null;
                        $composite = false;
                        $maxCharBuffer = null;
                    } else {
                        throw new InvalidArgumentException("Failed to expand token, invalid at col: $i");
                    }
                    break;
                case ',':
                    if ($token !== null) {
                        if ($maxCharBuffer !== null && $maxCharBuffer === '') {
                            throw new InvalidArgumentException("Found an empty prefix at col: $i");
                        }
                        $expanded = self::expandToken($operator, $token, $composite, self::getMaxChar($maxCharBuffer, $i), $firstToken, $substitutions, $result, $i);
                        if ($expanded && $firstToken) {
                            $firstToken = false;
                        }
                        $token = '';
                        $composite = false;
                        $maxCharBuffer = null;
                        break;
                    }
                    // Intentional fall-through for commas outside the {}
                default:
                    if ($token !== null) {
                        if ($operator === null) {
                            $operator = self::getOperator($character, $token, $i);
                        } elseif ($maxCharBuffer !== null) {
                            if (is_numeric($character)) {
                                $maxCharBuffer .= $character;
                            } else {
                                throw new InvalidArgumentException("Illegal character identified in the token at col: $i");
                            }
                        } else {
                            if ($character === ':') {
                                $maxCharBuffer = '';
                            } elseif ($character === '*') {
                                $composite = true;
                            } else {
                                self::validateLiteral($character, $i);
                                $token .= $character;
                            }
                        }
                    } else {
                        if (ord($character) > 0x7F) {
                            $result .= sprintf("%%%02X", ord($character));
                        } else {
                            $result .= $character;
                        }
                    }
                    break;
            }
        }

        if ($token === null) {
            return $result;
        } else {
            throw new InvalidArgumentException("Unterminated token");
        }
    }

    /**
     * @param string $op
     * @param string &$result
     */
    private static function addPrefix(string $op, string &$result): void {
        switch ($op) {
            case 'HASH':
                $result .= '#';
                break;
            case 'DOT':
                $result .= '.';
                break;
            case 'SLASH':
                $result .= '/';
                break;
            case 'SEMICOLON':
                $result .= ';';
                break;
            case 'QUESTION_MARK':
                $result .= '?';
                break;
            case 'AMP':
                $result .= '&';
                break;
            default:
                return;
        }
    }

    /**
     * @param string $op
     * @param string &$result
     */
    private static function addSeparator(string $op, string &$result): void {
        switch ($op) {
            case 'DOT':
                $result .= '.';
                break;
            case 'SLASH':
                $result .= '/';
                break;
            case 'SEMICOLON':
                $result .= ';';
                break;
            case 'QUESTION_MARK':
            case 'AMP':
                $result .= '&';
                break;
            default:
                $result .= ',';
                return;
        }
    }

    /**
     * @param string $op
     * @param string $token
     * @param mixed $value
     * @param string &$result
     * @param int $maxChar
     */
    private static function addValue(string $op, string $token, $value, string &$result, int $maxChar): void {
        switch ($op) {
            case 'PLUS':
            case 'HASH':
                self::addExpandedValue(null, $value, $result, $maxChar, false);
                break;
            case 'QUESTION_MARK':
            case 'AMP':

                $result .= $token . '=';
                self::addExpandedValue(null, $value, $result, $maxChar, true);
                break;
            case 'SEMICOLON':
                $result .= $token;
                self::addExpandedValue("=", $value, $result, $maxChar, true);
                break;
            case 'DOT':
            case 'SLASH':
            case 'NO_OP':
                self::addExpandedValue(null, $value, $result, $maxChar, true);
        }
    }

    /**
     * @param string $op
     * @param string $token
     * @param mixed $value
     * @param string &$result
     * @param int $maxChar
     */
    private static function addValueElement(string $op, string $token, $value, string &$result, int $maxChar): void {
        switch ($op) {
            case 'PLUS':
            case 'HASH':
                self::addExpandedValue(null, $value, $result, $maxChar, false);
                break;
            case 'QUESTION_MARK':
            case 'AMP':
            case 'SEMICOLON':
            case 'DOT':
            case 'SLASH':
            case 'NO_OP':
                self::addExpandedValue(null, $value, $result, $maxChar, true);
        }
    }

    /**
     * @param string $cp
     */
    private static function isSurrogate(string $cp) {
        if (empty($cp)) {
            return true;
        }
        $codePoint = mb_ord($cp, 'UTF-8');
        return ($codePoint >= 0xD800 && $codePoint <= 0xDFFF);
    }
    
    /**
     * @param string $cp
     */
    private static function isIprivate(string $cp) {
        if (empty($cp)) {
            return false;
        }
        $codePoint = mb_ord($cp, 'UTF-8');
        return (0xE000 <= $codePoint && $codePoint <= 0xF8FF);
    }
    
    /**
     * @param string $cp
     */
    private static function isUcschar(string $cp) {
        if (empty($cp)) {
            return false;
        }
        $codePoint = mb_ord($cp, 'UTF-8');
        return (0xA0 <= $codePoint && $codePoint <= 0xD7FF)
            || (0xF900 <= $codePoint && $codePoint <= 0xFDCF)
            || (0xFDF0 <= $codePoint && $codePoint <= 0xFFEF);
    }

    /**
     * @param mixed $prefix
     * @param mixed $value
     * @param string &$result
     * @param int $maxChar
     * @param bool $replaceReserved
     */
    private static function addExpandedValue($prefix, $value, string &$result, int $maxChar, bool $replaceReserved): void {
        $stringValue = self::convertNativeTypes($value);
        $charLen = mb_strlen($stringValue, 'UTF-8');
        $max = ($maxChar !== -1) ? min($maxChar, $charLen) : $charLen;
        $result .= '';
        $reservedBuffer = null;

        if ($max > 0 && $prefix !== null) {
            $result .= $prefix;
        }

        for ($i = 0; $i < $max; $i++) {
            $character = mb_substr($stringValue, $i, 1, 'UTF-8');

            if ($character === '%' && !$replaceReserved) {
                $reservedBuffer = '';
            }

            $toAppend = $character;
            $codePoint = mb_ord($character, 'UTF-8');
            if ($codePoint !== false && $codePoint > 0x7F || self::isSurrogate($character) || $replaceReserved || self::isUcschar($character) || self::isIprivate($character)) {
                $toAppend = urlencode($toAppend);
            }

            if ($reservedBuffer !== null) {
                $reservedBuffer .= $toAppend;

                if (strlen($reservedBuffer) === 3) {
                    $isEncoded = false;
                    try {
                        $decoded = urldecode($reservedBuffer);
                        $isEncoded = ($decoded !== $reservedBuffer);
                    } catch (Exception $e) {
                        // ignore
                    }

                    if ($isEncoded) {
                        $result .= $reservedBuffer;
                    } else {
                        $result .= "%25";
                        // only if !$replaceReserved
                        $result .= substr($reservedBuffer, 1);
                    }
                    $reservedBuffer = null;
                }
            } else {
                if ($character === ' ') {
                    $result .= "%20";
                } elseif ($character === '%') {
                    $result .= "%25";
                } else {
                    $result .= $toAppend;
                }
            }
        }

        if ($reservedBuffer !== null) {
            $result .= "%25";
            $result .= substr($reservedBuffer, 1);
        }
    }

    /**
     * @param mixed $value
     * @return bool
     */
    private static function isList($value): bool {
        return is_array($value);
    }

    /**
     * @param mixed $value
     * @return bool
     */
    private static function isMap($value): bool {
        if (is_array($value)) {
            // https://stackoverflow.com/a/173479/7898052
            if (!function_exists('array_is_list')) {
                if ($value === []) {
                    return true;
                }
                return array_keys($value) !== range(0, count($value) - 1);
            }
            return !array_is_list($value);
        }
        return false;
    }

    /**
     * @param mixed $value
     * @param int $col
     * @return string
     */
    private static function getSubstitutionType($value, int $col): string {
        if ($value === null) {
            return 'EMPTY';
        } elseif (self::isNativeType($value)) {
            return 'STRING';
        } elseif (self::isMap($value)) {
            return 'MAP';
        } elseif (self::isList($value)) {
            return 'LIST';
        } else {
            throw new InvalidArgumentException("Illegal class passed as substitution, found " . get_class($value) . " at col: $col");
        }
    }

    /**
     * @param string $substType
     * @param mixed $value
     * @return bool
     */
    private static function isEmpty(string $substType, $value): bool {
        if ($value === null) {
            return true;
        } else {
            switch ($substType) {
                case 'STRING':
                    return false;
                case 'LIST':
                case 'MAP':
                    return empty($value);
                default:
                    return true;
            }
        }
    }

    /**
     * @param mixed $value
     * @return bool
     */
    private static function isNativeType($value): bool {
        if (is_string($value) ||
            is_bool($value) ||
            is_int($value) ||
            is_float($value) ||
            is_double($value)) {
            return true;
        }
        return false;
    }

    /**
     * @param mixed $value
     * @return string
     */
    private static function convertNativeTypes($value): string {
        if (is_bool($value)) {
            if ($value) {
                return "true";
            } else {
                return "false";
            }
        } else if (is_string($value) || is_int($value) || is_float($value) || is_double($value)) {
            return (string)$value;
        } else {
            throw new InvalidArgumentException("Illegal class passed as substitution: $value");
        }
    }

    /**
     * @param string $operator
     * @param string $token
     * @param bool $composite
     * @param int $maxChar
     * @param bool $firstToken
     * @param array $substitutions
     * @param string &$result
     * @param int $col
     * @return bool
     */
    private static function expandToken(string $operator, string $token, bool $composite, int $maxChar, bool $firstToken, array $substitutions, string &$result, int $col): bool {
        if (empty($token)) {
            throw new InvalidArgumentException("Found an empty token at col: $col");
        }

        self::checkVarname($token, $col);

        $value = $substitutions[$token] ?? null;
        $substType = self::getSubstitutionType($value, $col);
        if ($substType === 'EMPTY' || self::isEmpty($substType, $value)) {
            return false;
        }

        if ($firstToken) {
            self::addPrefix($operator, $result);
        } else {
            self::addSeparator($operator, $result);
        }

        switch ($substType) {
            case 'STRING':
                self::addStringValue($operator, $token, $value, $result, $maxChar);
                break;
            case 'LIST':
                self::addListValue($operator, $token, $value, $result, $maxChar, $composite);
                break;
            case 'MAP':
                self::addMapValue($operator, $token, $value, $result, $maxChar, $composite);
                break;
        }

        return true;
    }

    /**
     * @param string $operator
     * @param string $token
     * @param mixed $value
     * @param string $result
     * @param int $maxChar
     */
    private static function addStringValue(string $operator, string $token, $value, string &$result, int $maxChar) {
        self::addValue($operator, $token, $value, $result, $maxChar);
    }

    /**
     * @param string $operator
     * @param string $token
     * @param array<mixed> $value
     * @param string &$result
     * @param int $maxChar
     * @param bool $composite
     * @return bool
     */
    private static function addListValue(string $operator, string $token, array $value, string &$result, int $maxChar, bool $composite): bool {
        $first = true;
        foreach ($value as $v) {
            if ($first) {
                self::addValue($operator, $token, $v, $result, $maxChar);
                $first = false;
            } else {
                if ($composite) {
                    self::addSeparator($operator, $result);
                    self::addValue($operator, $token, $v, $result, $maxChar);
                } else {
                    $result .= ',';
                    self::addValueElement($operator, $token, $v, $result, $maxChar);
                }
            }
        }
        return !$first;
    }

    /**
     * @param string $operator
     * @param string $token
     * @param array<mixed> $value
     * @param string &$result
     * @param int $maxChar
     * @param bool $composite
     * @return bool
     */
    private static function addMapValue(string $operator, string $token, array $value, string &$result, int $maxChar, bool $composite): bool {
        $first = true;
        if ($maxChar !== -1) {
            throw new InvalidArgumentException("Value trimming is not allowed on Maps");
        }
        foreach ($value as $k => $v) {
            if ($composite) {
                if (!$first) {
                    self::addSeparator($operator, $result);
                }
                self::addValueElement($operator, $token, (string)$k, $result, $maxChar);
                $result .= '=';
            } else {
                if ($first) {
                    self::addValue($operator, $token, (string)$k, $result, $maxChar);
                } else {
                    $result .= ',';
                    self::addValueElement($operator, $token, (string)$k, $result, $maxChar);
                }
                $result .= ',';
            }
            self::addValueElement($operator, $token, $v, $result, $maxChar);
            $first = false;
        }
        return !$first;
    }

}

File: dart/lib/std_uritemplate.dart
// ignore_for_file: constant_identifier_names

import 'dart:math';

enum _SubstitutionType { EMPTY, STRING, LIST, MAP }

enum _Operator { PLUS, HASH, DOT, SLASH, SEMICOLON, QUESTION_MARK, AMP, NO_OP }

extension on DateTime {
  /// This is the same as [toIso8601String] but without milli- and microseconds.
  String toIso8601StringWithoutMilliseconds() {
    String y =
        (year >= -9999 && year <= 9999) ? _fourDigits(year) : _sixDigits(year);
    String m = _twoDigits(month);
    String d = _twoDigits(day);
    String h = _twoDigits(hour);
    String min = _twoDigits(minute);
    String sec = _twoDigits(second);
    if (isUtc) {
      return "$y-$m-${d}T$h:$min:${sec}Z";
    } else {
      return "$y-$m-${d}T$h:$min:$sec";
    }
  }

  // Function below are copied from dart:core DateTime

  static String _fourDigits(int n) {
    int absN = n.abs();
    String sign = n < 0 ? "-" : "";
    if (absN >= 1000) return "$n";
    if (absN >= 100) return "${sign}0$absN";
    if (absN >= 10) return "${sign}00$absN";
    return "${sign}000$absN";
  }

  static String _sixDigits(int n) {
    assert(n < -9999 || n > 9999);
    int absN = n.abs();
    String sign = n < 0 ? "-" : "+";
    if (absN >= 100000) return "$sign$absN";
    return "${sign}0$absN";
  }

  static String _twoDigits(int n) {
    if (n >= 10) return "$n";
    return "0$n";
  }
}

class StdUriTemplate {
  const StdUriTemplate._();

  static String expand(String template, Map<String, Object?> substitutions) {
    return _expandImpl(template, substitutions);
  }

  static void _validateLiteral(String c, int col) {
    switch (c) {
      case '+':
      case '#':
      case '/':
      case ';':
      case '?':
      case '&':
      case ' ':
      case '!':
      case '=':
      case '\$':
      case '|':
      case '*':
      case ':':
      case '~':
      case '-':
        throw ArgumentError(
            "Illegal character identified in the token at col: $col ($c)");
      default:
        break;
    }
  }

  static int _getMaxChar(StringBuffer buffer, int col) {
    if (buffer.isEmpty) {
      return -1;
    } else {
      return int.tryParse(buffer.toString()) ??
          (throw ArgumentError("Cannot parse max chars at col: $col"));
    }
  }

  static _Operator _getOperator(String c, StringBuffer token, int col) {
    switch (c) {
      case '+':
        return _Operator.PLUS;
      case '#':
        return _Operator.HASH;
      case '.':
        return _Operator.DOT;
      case '/':
        return _Operator.SLASH;
      case ';':
        return _Operator.SEMICOLON;
      case '?':
        return _Operator.QUESTION_MARK;
      case '&':
        return _Operator.AMP;
      default:
        _validateLiteral(c, col);
        token.write(c);
        return _Operator.NO_OP;
    }
  }

  static String _expandImpl(String str, Map<String, Object?> substitutions) {
    final result = StringBuffer();

    var toToken = false;
    final token = StringBuffer();

    _Operator? operator;
    var composite = false;
    var toMaxCharBuffer = false;

    final maxCharBuffer = StringBuffer();
    var firstToken = true;

    for (var i = 0; i < str.length; i++) {
      final character = str[i];
      switch (character) {
        case '{':
          toToken = true;
          token.clear();
          firstToken = true;
          break;
        case '}':
          if (toToken) {
            assert(operator != null, "Operator cannot be null");

            final expanded = _expandToken(
              operator!,
              token.toString(),
              composite,
              _getMaxChar(maxCharBuffer, i),
              firstToken,
              substitutions,
              result,
              i,
            );
            if (expanded && firstToken) {
              firstToken = false;
            }
            toToken = false;
            token.clear();
            operator = null;
            composite = false;
            toMaxCharBuffer = false;
            maxCharBuffer.clear();
          } else {
            throw ArgumentError("Failed to expand token, invalid at col: $i");
          }
          break;
        case ',':
          if (toToken) {
            assert(operator != null, "Operator cannot be null");

            final expanded = _expandToken(
              operator!,
              token.toString(),
              composite,
              _getMaxChar(maxCharBuffer, i),
              firstToken,
              substitutions,
              result,
              i,
            );
            if (expanded && firstToken) {
              firstToken = false;
            }
            token.clear();
            composite = false;
            toMaxCharBuffer = false;
            maxCharBuffer.clear();
            break;
          }
          // Intentional fall-through for commas outside the {}
          continue;
        default:
          if (toToken) {
            if (operator == null) {
              operator = _getOperator(character, token, i);
            } else if (toMaxCharBuffer) {
              if (int.tryParse(character) != null) {
                maxCharBuffer.write(character);
              } else {
                throw ArgumentError(
                  "Illegal character identified in the token at col: $i",
                );
              }
            } else {
              if (character == ':') {
                toMaxCharBuffer = true;
                maxCharBuffer.clear();
              } else if (character == '*') {
                composite = true;
              } else {
                _validateLiteral(character, i);
                token.write(character);
              }
            }
          } else {
            result.write(character);
          }
          break;
      }
    }

    if (!toToken) {
      return result.toString();
    } else {
      throw ArgumentError("Unterminated token");
    }
  }

  static void _addPrefix(_Operator op, StringBuffer result) {
    switch (op) {
      case _Operator.HASH:
        result.write('#');
        break;
      case _Operator.DOT:
        result.write('.');
        break;
      case _Operator.SLASH:
        result.write('/');
        break;
      case _Operator.SEMICOLON:
        result.write(';');
        break;
      case _Operator.QUESTION_MARK:
        result.write('?');
        break;
      case _Operator.AMP:
        result.write('&');
        break;
      default:
        return;
    }
  }

  static void _addSeparator(_Operator op, StringBuffer result) {
    switch (op) {
      case _Operator.DOT:
        result.write('.');
        break;
      case _Operator.SLASH:
        result.write('/');
        break;
      case _Operator.SEMICOLON:
        result.write(';');
        break;
      case _Operator.QUESTION_MARK:
      case _Operator.AMP:
        result.write('&');
        break;
      default:
        result.write(',');
        return;
    }
  }

  static void _addValue(_Operator op, String token, Object? value,
      StringBuffer result, int maxChar) {
    switch (op) {
      case _Operator.PLUS:
      case _Operator.HASH:
        _addExpandedValue(null, value, result, maxChar, false);
        break;
      case _Operator.QUESTION_MARK:
      case _Operator.AMP:
        result.write('$token=');
        _addExpandedValue(null, value, result, maxChar, true);
        break;
      case _Operator.SEMICOLON:
        result.write(token);
        _addExpandedValue('=', value, result, maxChar, true);
        break;
      case _Operator.DOT:
      case _Operator.SLASH:
      case _Operator.NO_OP:
        _addExpandedValue(null, value, result, maxChar, true);
    }
  }

  static void _addValueElement(_Operator op, String token, Object? value,
      StringBuffer result, int maxChar) {
    switch (op) {
      case _Operator.PLUS:
      case _Operator.HASH:
        _addExpandedValue(null, value, result, maxChar, false);
        break;
      case _Operator.QUESTION_MARK:
      case _Operator.AMP:
      case _Operator.SEMICOLON:
      case _Operator.DOT:
      case _Operator.SLASH:
      case _Operator.NO_OP:
        _addExpandedValue(null, value, result, maxChar, true);
    }
  }

  static bool _isSurrogate(String cp) {
    assert(cp.isNotEmpty);
    final codePoint = cp.codeUnitAt(0);
    return (codePoint >= 0xD800 && codePoint <= 0xDFFF);
  }

  static bool _isIprivate(String cp) {
    assert(cp.isNotEmpty);
    final codePoint = cp.codeUnitAt(0);
    return (0xE000 <= codePoint && codePoint <= 0xF8FF);
  }

  static bool _isUcschar(String cp) {
    assert(cp.isNotEmpty);
    final codePoint = cp.codeUnitAt(0);
    return (0xA0 <= codePoint && codePoint <= 0xD7FF) ||
        (0xF900 <= codePoint && codePoint <= 0xFDCF) ||
        (0xFDF0 <= codePoint && codePoint <= 0xFFEF);
  }

  static void _addExpandedValue(
    String? prefix,
    Object? value,
    StringBuffer result,
    int maxChar,
    bool replaceReserved,
  ) {
    final stringValue = _convertNativeTypes(value);
    final runes = stringValue.runes;
    final max = (maxChar != -1) ? min(maxChar, runes.length) : runes.length;

    var toReserved = false;
    final reservedBuffer = StringBuffer();

    if (max > 0 && prefix != null) {
      result.write(prefix);
    }

    for (var i = 0; i < max; i++) {
      final character = String.fromCharCode(runes.elementAt(i));

      if (character == '%' && !replaceReserved) {
        toReserved = true;
        reservedBuffer.clear();
      }

      var toAppend = character;
      if (replaceReserved ||
          _isSurrogate(character) ||
          _isUcschar(character) ||
          _isIprivate(character)) {
        toAppend = Uri.encodeQueryComponent(toAppend);
      }

      if (toReserved) {
        reservedBuffer.write(toAppend);

        if (reservedBuffer.length == 3) {
          var isEncoded = false;
          try {
            final decoded = Uri.decodeComponent(reservedBuffer.toString());
            isEncoded = (decoded != reservedBuffer.toString());
          } catch (e) {
            // ignore
          }

          if (isEncoded) {
            result.write(reservedBuffer);
          } else {
            result.write("%25");
            // only if !replaceReserved
            result.write(reservedBuffer.toString().substring(1));
          }
          toReserved = false;
          reservedBuffer.clear();
        }
      } else {
        if (character == ' ') {
          result.write("%20");
        } else if (character == '%') {
          result.write("%25");
        } else {
          result.write(toAppend);
        }
      }
    }

    if (toReserved) {
      result.write("%25");
      result.write(reservedBuffer.toString().substring(1));
    }
  }

  static bool _isList(Object value) {
    return value is Iterable;
  }

  static bool _isMap(Object value) {
    return value is Map;
  }

  static _SubstitutionType _getSubstitutionType(Object? value, int col) {
    if (value == null) {
      return _SubstitutionType.EMPTY;
    } else if (_isNativeType(value)) {
      return _SubstitutionType.STRING;
    } else if (_isMap(value)) {
      return _SubstitutionType.MAP;
    } else if (_isList(value)) {
      return _SubstitutionType.LIST;
    } else {
      throw ArgumentError(
        "Illegal class passed as substitution, found ${value.runtimeType}"
        " at col: $col",
      );
    }
  }

  static bool _isEmpty(_SubstitutionType substType, Object? value) {
    if (value == null) {
      return true;
    } else {
      return switch (substType) {
        _SubstitutionType.STRING => false,
        _SubstitutionType.LIST => value is Iterable
            ? value.isEmpty
            : throw UnimplementedError(
                'Impossible case of _SubstitutionType being list'
                ' and value not being an Iterable.',
              ),
        _SubstitutionType.MAP => value is Map
            ? value.isEmpty
            : throw UnimplementedError(
                'Impossible case of _SubstitutionType being list'
                ' and value not being an Iterable.',
              ),
        _ => true
      };
    }
  }

  static bool _isNativeType(Object? value) {
    if (value is String || value is bool || value is num || value is DateTime) {
      return true;
    }
    return false;
  }

  static String _convertNativeTypes(Object? value) {
    if (value is bool || value is String || value is num) {
      return value.toString();
    } else if (value is DateTime) {
      return value.toUtc().toIso8601StringWithoutMilliseconds();
    } else {
      return '';
    }
  }

  static bool _expandToken(
    _Operator operator,
    String token,
    bool composite,
    int maxChar,
    bool firstToken,
    Map<String, Object?> substitutions,
    StringBuffer result,
    int col,
  ) {
    if (token.isEmpty) {
      throw ArgumentError("Found an empty token at col: $col");
    }

    final value = substitutions[token];
    final substType = _getSubstitutionType(value, col);
    if (substType == _SubstitutionType.EMPTY || _isEmpty(substType, value)) {
      return false;
    }

    if (firstToken) {
      _addPrefix(operator, result);
    } else {
      _addSeparator(operator, result);
    }

    switch (substType) {
      case _SubstitutionType.STRING:
        _addStringValue(operator, token, value, result, maxChar);
        break;
      case _SubstitutionType.LIST:
        _addListValue(
          operator,
          token,
          value as Iterable,
          result,
          maxChar,
          composite,
        );
        break;
      case _SubstitutionType.MAP:
        _addMapValue(
          operator,
          token,
          (value as Map).cast<String, Object?>(),
          result,
          maxChar,
          composite,
        );
        break;
      default:
        // do nothing
        break;
    }

    return true;
  }

  static void _addStringValue(
    _Operator operator,
    String token,
    Object? value,
    StringBuffer result,
    int maxChar,
  ) {
    _addValue(operator, token, value, result, maxChar);
  }

  static bool _addListValue(
    _Operator operator,
    String token,
    Iterable<Object?> value,
    StringBuffer result,
    int maxChar,
    bool composite,
  ) {
    var first = true;
    for (final v in value) {
      if (first) {
        _addValue(operator, token, v, result, maxChar);
        first = false;
      } else {
        if (composite) {
          _addSeparator(operator, result);
          _addValue(operator, token, v, result, maxChar);
        } else {
          result.write(',');
          _addValueElement(operator, token, v, result, maxChar);
        }
      }
    }
    return !first;
  }

  static bool _addMapValue(
    _Operator operator,
    String token,
    Map<String, Object?> value,
    StringBuffer result,
    int maxChar,
    bool composite,
  ) {
    var first = true;
    if (maxChar != -1) {
      throw ArgumentError("Value trimming is not allowed on Maps");
    }
    for (var MapEntry(key: k, value: v) in value.entries) {
      if (composite) {
        if (!first) {
          _addSeparator(operator, result);
        }
        _addValueElement(operator, token, k, result, maxChar);
        result.write('=');
      } else {
        if (first) {
          _addValue(operator, token, k, result, maxChar);
        } else {
          result.write(',');
          _addValueElement(operator, token, k, result, maxChar);
        }
        result.write(',');
      }
      _addValueElement(operator, token, v, result, maxChar);
      first = false;
    }
    return !first;
  }
}
